### DISCUSSION TITLE
request help:  The X-Forwarded-For header of the request to the upstream cannot be changed

### DISCUSSION THREAD

**Discussion by jagerzhang** [apache/apisix#5091] [Q&A] [answered]
### Issue description

For historical reasons, our upstream service uses IP authentication, the real client IP is get from x-forwarded-for header. 
Now it needs to be changed to gateway HMAC authentication, so the upstream IP authentication needs to be shielded through the gateway.

I tried to modify x-forwarded-for through proxy rewirte plug-in, but it didn't take effect:
![image](https://user-images.githubusercontent.com/9711651/133709684-87f2e81a-1d32-4032-8e80-4da4b2d886a1.png)

when using the Kong gateway, we changed x-forwarded-for through serverless:
![image](https://user-images.githubusercontent.com/9711651/133709759-5f2435f5-41bf-40b6-a574-394c15e7404c.png)

However,  I tried in APISIX, but it does not take effect:

attempt 1
```
"serverless-pre-function": {
      "functions": [
        "return function(conf, ctx) ngx.req.set_header('X-Forwarded-For', 2.2.2.2) end"
      ],
      "phase": "balancer"
    }
```

attempt 2
```
"serverless-pre-function": {
      "functions": [
        "return function(ctx)    ctx.upstream_x_forwarded_for = '2.2.2.2'; end"
      ],
      "phase": "balancer"
    }
```

attempt 3
```
"serverless-pre-function": {
      "functions": [
        "return function(conf, ctx) local core = require('apisix.core');core.request.set_header(ctx, 'X-Forwarded-For', '2.2.2.2'); end"
      ],
      "phase": "balancer"
    }
```

Please take a look.

### Environment

- apisix version (cmd: `apisix version`): 2.9
- OS (cmd: `uname -a`): Docker image
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API): 3.x
- apisix-dashboard version, if have: 2.8

**Comment by spacewander** [comment]
Currently, we always override the XFF in https://github.com/apache/apisix/blob/c907ef0f663eba5f10686e13e431bd007e7c5ac3/apisix/cli/ngx_tpl.lua#L597.

We can handle it like X-Forwarded-Proto in https://github.com/apache/apisix/blob/c907ef0f663eba5f10686e13e431bd007e7c5ac3/apisix/init.lua#L271

PR is welcome!

**Comment by jagerzhang** [comment]
> Currently, we always override the XFF in
> 
> https://github.com/apache/apisix/blob/c907ef0f663eba5f10686e13e431bd007e7c5ac3/apisix/cli/ngx_tpl.lua#L597
> 
> .
> We can handle it like X-Forwarded-Proto in
> 
> https://github.com/apache/apisix/blob/c907ef0f663eba5f10686e13e431bd007e7c5ac3/apisix/init.lua#L271
> 
> PR is welcome!

Do we need to write a plug-in ourselves?

**Comment by shuaijinchao** [comment, marked as answer]
Can be handled in this way temporarily

```bash
curl -i http://127.0.0.1:9080/apisix/admin/routes/10  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/test/index.html",
    "plugins": {
        "serverless-pre-function": {
            "phase": "rewrite",
            "functions": [
                "return function(conf, ctx) ngx.var.var_x_forwarded_for = tostring(\"1.1.1.1\") end"
            ]
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    }
}'
```
---
![image](https://user-images.githubusercontent.com/8529452/133728294-61d6f5ac-6ecb-470b-be9c-7646715ac590.png)

**Comment by jagerzhang** [comment]
> Can be handled in this way temporarily
> 
> ```shell
> curl -i http://127.0.0.1:9080/apisix/admin/routes/10  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
> {
>     "uri": "/test/index.html",
>     "plugins": {
>         "serverless-pre-function": {
>             "phase": "rewrite",
>             "functions": [
>                 "return function(conf, ctx) ngx.var.var_x_forwarded_for = tostring(\"1.1.1.1\") end"
>             ]
>         }
>     },
>     "upstream": {
>         "type": "roundrobin",
>         "nodes": {
>             "127.0.0.1:1980": 1
>         }
>     }
> }'
> ```
> 
> ![image](https://user-images.githubusercontent.com/8529452/133728294-61d6f5ac-6ecb-470b-be9c-7646715ac590.png)

@shuaijinchao 
it works for me, tks! 

There is only one difference in variable name between Kong gateway and APISIX gateway:
`ngx.var.upstream_x_forwarded_for  ==>  ngx.var.var_x_forwarded_for`

**Reply by juzn01** [reply]
Hi bro, if i want to remove x_forwarded_for header , how i can do ??
thank you.

**Reply by bk13699** [reply]
36556a37d1264f88cbfdf5f54a368f641e8b65f5

**Reply by bk13699** [reply]
curl -i http://127.0.0.1:9080/apisix/admin/routes/10  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/test/index.html",
    "plugins": {
        "serverless-pre-function": {
            "phase": "rewrite",
            "functions": [
                "return function(conf, ctx) ngx.var.var_x_forwarded_for = tostring(\"1.1.1.1\") end"
            ]
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    }
}'

**Reply by bk13699** [reply]
curl -i http://127.0.0.1:9080/apisix/admin/routes/10  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/test/index.html",
    "plugins": {
        "serverless-pre-function": {
            "phase": "rewrite",
            "functions": [
                "return function(conf, ctx) ngx.var.var_x_forwarded_for = tostring(\"1.1.1.1\") end"
            ]
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:1980": 1
        }
    }
}'
